dm/docs/RFCS/20211130_enhanced_pre_checker.md

```diff
@@ -0,0 +1,82 @@
+# Enhanced pre-check Design
+
+## Background
+
+Before the DM’s task starts, we should check some items to avoid start-task failures. You can see the details in our [document](https://docs.pingcap.com/zh/tidb-data-migration/stable/precheck#%E5%85%B3%E9%97%AD%E6%A3%80%E6%9F%A5%E9%A1%B9). In order to allow some users to use DM normally under certain circumstances, we can also set ignore-check-items in task config to ignore some items that you don’t want to check. Now, we find some shortcomings in regard to check-items.
+
+### Bad user habits
+
+We allow users to ignore all check-items, in which case the user's authority is too large to perform unexpected operations. 
+
+### Too much time overhead
+
+If we have a large number of tables in source, we will take too much time in checking table schema, sharding table consistency and sharding table auto increment key.
+
+### Inadequate check
+
+* If downstream creates tables manually and the new downstream’s auto increment ID is not the same as the upstream, we shouldn’t check **auto_increment_ID** for errors. Users should be responsible for what they set.
+* Dump privilege only checks RELOAD and SELECT. However, Dumpling supports different [consistency configurations](https://docs.pingcap.com/zh/tidb/stable/dumpling-overview#%E8%B0%83%E6%95%B4-dumpling-%E7%9A%84%E6%95%B0%E6%8D%AE%E4%B8%80%E8%87%B4%E6%80%A7%E9%80%89%E9%A1%B9), which need more privilege.
+* If online-ddl is set by true and a DDL is in online-ddl stage, DM will have a problem in all mode. Specifically, ghost table has been created, is executing the DDL, but is not renamed yet. In this case, DM will report an error when the ghost table is renamed after the dump phase. You can learn more about online-ddl [here](https://docs.pingcap.com/zh/tidb-data-migration/stable/feature-online-ddl).
+* For schema_of_shard_tables, whatever pessimistic task and optimistic task, we all check it by comparing all sharding tables’ structures for consistency simply. For optimistic mode, we can do better.
+
+## Proposal
+
+### Speed ​​up check
+
+1. Support concurrent check
+    - table_schema
+    - schema_of_shard_tables
+    - auto_increment_ID
+2. Use mydumper.threads as **source_connection_concurrency**, which should update in our document.
+
+#### How to speed up?
+
+Since every checker is concurrent, we can split tables to **source_connection_concurrency** part, and create a checker for every part. 
+
+### Optimize some check
+
+1. We needn’t check **auto_increment_ID**, if the column of auto increment ID in upstream does not has an unique constraint in downstream.
```

Comment:
Please add a description that the **auto_increment_ID** is only checked for sharding tables.

Reply:
Added in background https://github.com/pingcap/ticdc/pull/3674/commits/eabe327b15bbd945a4b80d659237202b22f39433

Reply:
Seems that it is difficult to support concurrent check auto_increment_ID. Now we check it by `mapping` which is deprecated. If we don't set the `mapping` and don't ignore the `auto_increment_ID`, the pre-check will report an error.

So how about we report a warning to user and tell them the method that can resolve the PK/UK conflict. I think this is enough.

also cc @lance6716

Reply:
sgtm.

Reply:
update in https://github.com/pingcap/ticdc/pull/3674/commits/66d1ebd59dc18ead627a8d9fd89df3a4f2ac7fde